crates/ethrpc/src/buffered.rs

```diff
@@ -293,12 +251,191 @@ where
+///     ),
+/// ];
+/// let trace_ids = vec![Some("1001".to_string()), None, Some("1002".to_string())];
+/// let metadata_header = build_rpc_metadata_header(requests, trace_ids);
+/// assert_eq!(
+///     metadata_header,
+///     "null:eth_sendTransaction(1)|1001:eth_call(0)|1002:eth_call(2)"
+/// );
+/// ```
+fn build_rpc_metadata_header<'a>(
+    requests: impl Iterator<Item = &'a (RequestId, Call)>,
+    trace_ids: impl Iterator<Item = &'a Option<String>>,
+) -> String {
+    let mut grouped_metadata: BTreeMap<String, BTreeMap<String, BTreeSet<usize>>> = BTreeMap::new();
+    for (idx, ((_, call), trace_id)) in requests.zip(trace_ids).enumerate() {
+        if let Call::MethodCall(call) = call {
+            let trace_id = trace_id.clone().unwrap_or("null".to_string());
+            grouped_metadata
+                .entry(trace_id)
+                .or_default()
+                .entry(call.method.clone())
+                .or_default()
+                .insert(idx);
+        }
+    }
+
+    let mut metadata_str = String::new();
+
+    let mut grouped_metadata_iter = grouped_metadata.into_iter().peekable();
+    while let Some((trace_id, methods)) = grouped_metadata_iter.next() {
+        // New entry starts with the trace_id
+        metadata_str.push_str(&format!("{}:", trace_id));
+
+        // Followed by the method names and their indices
+        let mut methods_iter = methods.into_iter().peekable();
+        while let Some((method, indices)) = methods_iter.next() {
+            metadata_str.push_str(&format!("{}(", method));
+
+            let indices_str = format_indices_as_ranges(indices);
+            metadata_str.push_str(&indices_str);
+
+            metadata_str.push(')');
+
+            if methods_iter.peek().is_some() {
+                metadata_str.push(',');
+            }
+        }
+
+        if grouped_metadata_iter.peek().is_some() {
+            metadata_str.push('|');
+        }
+    }
+
+    metadata_str
+}
+
+/// Formats a set of indices as a string of ranges.
+///
+/// This function takes a set of indices and formats them as a string where
+/// consecutive indices are represented as ranges. For example, the set
+/// `{1, 2, 3, 5, 6, 8}` would be formatted as the string `"1..3,5..6,8"`.
+///
+/// # Arguments
+///
+/// * `indices` - A set of indices to format. The indices should be unique and
+///   sorted in ascending order.
+///
+/// # Returns
+///
+/// This function returns a string representing the indices as ranges. Each
+/// range is formatted as `start..end`, and ranges are separated by commas.
+/// Single indices (i.e., indices that are not part of a range) are represented
+/// as themselves.
+///
+/// # Examples
+///
+/// ```
+/// let indices = vec![1, 2, 3, 5, 6, 8].into_iter().collect();
+/// let ranges = format_indices_as_ranges(indices);
+/// assert_eq!(ranges, "1..3,5..6,8");
+/// ```
+fn format_indices_as_ranges(indices: BTreeSet<usize>) -> String {
+    let indices = indices.into_iter().collect_vec();
```

Comment:
Collecting into a vector can be avoided here by converting to a plain iterator instead.
```rust
    let mut indices = indices.into_iter();
    let start = match indices.next() {
        Some(index) => index,
        None => return String::new();
    };
```